designs/warmreplicas.md

```diff
@@ -0,0 +1,131 @@
+extension watches) do not start until they have won leader election. This implies guaranteed
+downtime as even after leader election, the controller has to wait for the initial list to be served
+before it can start reconciling.
+
+## Proposal
+A warm replica is a replica with a queue pre-filled by sources started even when leader election is
+not won so that it is ready to start processing items as soon as the leader election is won. This
+proposal aims to add support for warm replicas in controller-runtime.
+
+### Context
+Mostly written to confirm my understanding, but also to provide context for the proposal.
+
+Controllers are a monolithic runnable with a `Start(ctx)` that
+1. Starts the watches [here](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.20.2/pkg/internal/controller/controller.go#L196-L213)
+2. Starts the workers [here](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.20.2/pkg/internal/controller/controller.go#L244-L252)
+There needs to be a way to decouple the two so that the watches can be started before the workers
+even as part of the same Runnable.
+
+If a runnable implements the `LeaderElectionRunnable` interface, the return value of the
+`NeedLeaderElection` function dictates whether or not it gets binned into the leader election
+runnables group [code](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.20.2/pkg/manager/runnable_group.go).
+
+Runnables in the leader election group are started only after the manager has won leader election,
+and all controllers are leader election runnables by default.
+
+### Design
+1. Add a new interface `WarmupRunnable` that allows for leader election runnables to specify
+behavior when manager is not in leader mode. This interface should be as follows:
+```go
+type WarmupRunnable interface {
+    Warmup(context.Context) error
+}
+```
+
+2. Controllers will implement this interface to specify behavior when the manager is not the leader.
+Add a new controller option `ShouldWarmupWithoutLeadership`. If set to true, then the main
+controller runnable will not start sources, and instead rely on the warmup runnable to start sources
+The option will be used as follows:
+```go
+type Controller struct {
+    // ...
+
+    // ShouldWarmupWithoutLeadership specifies whether the controller should start its sources
+    // when the manager is not the leader.
+    // Defaults to false, which means that the controller will wait for leader election to start
+    // before starting sources.
+    ShouldWarmupWithoutLeadership *bool
+
+    // ...
+}
+
+type runnableWrapper struct {
+    startFunc func (ctx context.Context) error
+}
+
+func(rw runnableWrapper) Start(ctx context.Context) error {
+    return rw.startFunc(ctx)
+}
+
+func (c *Controller[request]) Warmup(ctx context.Context) error {
+    if !c.ShouldWarmupWithoutLeadership {
+        return nil
+    }
+
+    return c.startEventSources(ctx)
```

Comment:
Lets maybe just  leave this out so we don't have to update the design doc everytime something changes?
```suggestion
    ...
```